### PR TITLE
[Zero-Dim] support 0-D output and 0-D as indice in __getitem__/__setitem__

### PR DESCRIPTION
**Pull request by zoooo0820**
### PR types

Others
### PR changes

APIs
### Describe

Pcard-66985

This PR is to support 0-D Tensor in Tensor `__getitem__`. From now, 0-D Tensor is semantically equivalent to **scalar** while 1-D Tensor is **vector**.

There will be two **incompatible behavior changes** in `__getitem__`:
- all axis has a scalar indice, the output will be a 0-D tensor.
```python
>>> x = paddle.ones((2,3,4))
>>> y = x[1,1,1]

# before this PR:
>>> y
Tensor(shape=[1], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       1.)

# after this PR:
>>> y
Tensor(shape=[], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       1.)
```

- In index element,  0-D Tensor will be treated as scalar, and 1-D Tensor will be treat as vector
```python
>>> x = paddle.ones((2,3,4))
>>> y = x[1,1,1]
>>> y
Tensor(shape=[], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       1.)

# is same as y = x[1,1,1]
>>> indice = paddle.ones((), dtype='int32')
>>> x[indice, indice, indice]
Tensor(shape=[], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       1.)

# behavior changes for 1-D Tensor
>>> x = paddle.ones((2,3,4))
>>> indice = paddle.ones([1], dtype='int32')

# before this PR, 1-D Tensor is treated as scalar
>>> x[indice].shape
[3, 4]

# after this PR, 1-D Tensor is treated as vector
>>> x[indice].shape
[1, 3, 4]
```

**NOTE** : Combine indexing is not supported now, and  1-d Tensor is treated as vector, so following usages are not supported, we will try to support these usages as soon as possible.
`x[1, paddle.ones([1], dtype='int32')]` / `x[:, paddle.ones([1], dtype='int32')]`

For `__setitem__` , now we can also use 0-D Tensor as indice
```python
>>> x = paddle.ones((2,3,2))
>>> indice = paddle.ones([], dtype='int32')
>>> x[indice, indice] = 10
>>> x
Tensor(shape=[2, 3, 2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[[1. , 1. ],
         [1. , 1. ],
         [1. , 1. ]],

        [[1. , 1. ],
         [10., 10.],
         [1. , 1. ]]])
```

0-D Tensor can also be used as a value to fill.
```python
>>> x = paddle.ones((2,3,2))
>>> indice = paddle.ones([], dtype='int32')
>>> value = paddle.ones([]) * 10
>>> x[indice, indice] = value
>>> x
Tensor(shape=[2, 3, 2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[[1. , 1. ],
         [1. , 1. ],
         [1. , 1. ]],

        [[1. , 1. ],
         [10., 10.],
         [1. , 1. ]]])
```

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/operators/dist_slice.py [hunk 1]
```diff
@@ -142,9 +140,7 @@ def update_dims_mapping(self, dist_op):
                 ref_indices.append(i)
 
         if ref_dims_mapping == []:
-            ref_dims_mapping = [-1]
             assert len(ref_dims_mapping) == len(out_dims_mapping)
-            assert ref_dims_mapping[0] == out_dims_mapping[0]
```
Comment (zhwesky2010) [inline python/paddle/distributed/auto_parallel/operators/dist_slice.py:147]: 这个感觉可以判断out_dims_mapping如果维度>0，就判断 assert ref_dims_mapping[0] == out_dims_mapping[0]，这样之前的检查逻辑也可以保留了

**Review by zhwesky2010 — DISMISSED** [review]

**Comment by zhwesky2010** [comment]
test_set_value_op_xpu、test_slice_op_xpu挂了

File: python/paddle/distributed/auto_parallel/operators/dist_slice.py (hunk 1 above)
Comment (zoooo0820) [reply python/paddle/distributed/auto_parallel/operators/dist_slice.py:147]: 这里特化了`ref_dims_mapping==[]`的场景，在下面else分支中会逐轴比对`ref_dims_mapping[i] / out_dims_mapping[i]`

**Review by zhwesky2010 — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM for paddle_enforce

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**